modules/sc-mesh-secure-deployment/src/2_0/features/cbma/auth/authServer.py

```diff
@@ -0,0 +1,136 @@
+        threading.Thread.__init__(self)
+        self.running = True
+        self.ipAddress = ip_address
+        self.port = port
+        self.CERT_PATH = cert_path
+        self.ca = ca_path
+        self.interface = interface
+        self.mymac = get_mac_addr(self.interface)
+        # Create the SSL context here and set it as an instance variable
+        self.context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+        self.context.verify_mode = ssl.CERT_REQUIRED
+        self.context.load_verify_locations(glob.glob(self.ca)[0])
+        self.context.load_cert_chain(
+            certfile=glob.glob(f"{self.CERT_PATH}/macsec_{self.mymac.replace(':', '')}.crt")[0],
+            keyfile=glob.glob(f"{self.CERT_PATH}/macsec_{self.mymac.replace(':', '')}.key")[0],
+        )
+        self.client_auth_results = {}
+        self.active_sockets = {}
+        self.client_auth_results_lock = threading.Lock()
+        self.active_sockets_lock = threading.Lock()
+        self.mua = mua
+
+    def handle_client(self, client_connection, client_address):
+        client_mac = extract_mac_from_ipv6(client_address[0])  # TODO: check if it is safe to do so
+        print("------------------server---------------------")
+        if client_mac not in self.mua.connected_peers_status:
+            with self.mua.connected_peers_status_lock:
+                self.mua.connected_peers_status[client_mac] = ["ongoing",0]  # Update status as ongoing, num of failed attempts = 0
+        else:
+            with self.mua.connected_peers_status_lock:
+                self.mua.connected_peers_status[client_mac][0] = "ongoing"  # Update status as ongoing, num of failed attempts = same as before
+        self.authenticate_client(client_connection, client_address, client_mac)
+
+    def authenticate_client(self, client_connection, client_address, client_mac):
+        secure_client_socket = self.context.wrap_socket(client_connection, server_side=True)
+        try:
+            client_cert = secure_client_socket.getpeercert(binary_form=True)
+            if not client_cert:
+                logger.error(f"Unable to get the certificate from the client {client_address[0]}", exc_info=True)
+                raise CertificateNoPresentError("Unable to get the certificate from the client")
+
+            auth = verify_cert(client_cert, self.ca, client_address[0], self.interface, logger)
+            with self.client_auth_results_lock:
+                self.client_auth_results[client_address[0]] = auth
+            if auth:
+                with self.active_sockets_lock:
+                    self.active_sockets[client_address[0]] = secure_client_socket
+                self.mua.auth_pass(secure_client_socket=secure_client_socket, client_mac=client_mac)
+            else:
+                # Handle the case when authentication fails, maybe send an error message
+                self.mua.auth_fail(client_mac=client_mac)
+                secure_client_socket.send(b"Authentication failed.")
```

Comment:
send() - don't guarantee complete packet send.

https://docs.python.org/3/howto/sockets.html#using-a-socket

You need to have known message sizes.  Use sendall() to send complete packet and on receiver side you need to know how long packet is coming.

Reply:
Great catch! But it shouldn't be possible to transmit data anyway if TLS verification failed, I imagine the other side will simply get a SSLError exception 🙂

Reply:
Doesn't justify wrong usage of socket.send()

Reply:
Sorry I didn't express myself very well, I meant there shouldn't be a send at all 😅

Reply:
--> remove socket.send() from here.